test/integration/hds_integration_test.cc

```diff
@@ -1043,5 +1050,124 @@ TEST_P(HdsIntegrationTest, SingleEndpointUnhealthyTlsMissingSocketMatch) {
+  hds_stream_->sendGrpcMessage(server_health_check_specifier_);
+  test_server_->waitForCounterGe("hds_delegate.requests", ++hds_requests_);
+
+  // Envoy sends health check messages to two endpoints
+  healthcheckEndpoints("cat");
+
+  // Endpoint responds to the health check
+  host_stream_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "404"}}, false);
+  host_stream_->encodeData(1024, true);
+  host2_stream_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, false);
+  host2_stream_->encodeData(1024, true);
+
+  // Receive updates until the one we expect arrives
+  ASSERT_TRUE(waitForClusterHealthResponse(envoy::config::core::v3::HEALTHY,
+                                           host2_upstream_->localAddress(), 1, 0, 0));
+
+  ASSERT_EQ(response_.endpoint_health_response().cluster_endpoints_health_size(), 2);
+
+  // store cluster response info for easier reference.
+  const auto& cluster_resp0 = response_.endpoint_health_response().cluster_endpoints_health(0);
+  const auto& cluster_resp1 = response_.endpoint_health_response().cluster_endpoints_health(1);
+
+  // check cluster info and sizes.
+  EXPECT_EQ(cluster_resp0.cluster_name(), "anna");
+  ASSERT_EQ(cluster_resp0.locality_endpoints_health_size(), 1);
+  EXPECT_EQ(cluster_resp1.cluster_name(), "cat");
+  ASSERT_EQ(cluster_resp1.locality_endpoints_health_size(), 1);
+
+  // store locality response info for easier reference.
+  const auto& locality_resp0 = cluster_resp0.locality_endpoints_health(0);
+  const auto& locality_resp1 = cluster_resp1.locality_endpoints_health(0);
+
+  // check locality info and sizes.
+  EXPECT_EQ(locality_resp0.locality().sub_zone(), "hobbiton");
+  ASSERT_EQ(locality_resp0.endpoints_health_size(), 1);
+  EXPECT_EQ(locality_resp1.locality().sub_zone(), "paris");
+  ASSERT_EQ(locality_resp1.endpoints_health_size(), 1);
+
+  // Check endpoints.
+  EXPECT_TRUE(checkEndpointHealthResponse(locality_resp0.endpoints_health(0),
+                                          envoy::config::core::v3::UNHEALTHY,
+                                          host_upstream_->localAddress()));
+
+  checkCounters(1, 2, 0, 1);
+  EXPECT_EQ(1, test_server_->counter("cluster.cat.health_check.success")->value());
+  EXPECT_EQ(0, test_server_->counter("cluster.cat.health_check.failure")->value());
+
+  // Create new specifier that removes the second cluster, and adds an endpoint to the first.
+  server_health_check_specifier_ =
+      makeHttpHealthCheckSpecifier(envoy::type::v3::CodecClientType::HTTP1, false);
+  Network::Utility::addressToProtobufAddress(
+      *host2_upstream_->localAddress(),
+      *server_health_check_specifier_.mutable_cluster_health_checks(0)
+           ->mutable_locality_endpoints(0)
+           ->add_endpoints()
+           ->mutable_address());
+
+  // Reset second endpoint for usage in our cluster.
+  AssertionResult result = host2_fake_connection_->close();
+  RELEASE_ASSERT(result, result.message());
```

Comment:
Optional nit: It would be good to be consistent about RELEASE_ASSERT and ASSERT_TRUE here.

Reply:
Thanks, I should be consistent here. changed to ASSERT_TRUEs.